--- device/oneplus/oneplus2/device.mk.old	2016-08-07 20:08:28.579247857 +1000
+++ device/oneplus/oneplus2/device.mk	2016-08-10 15:21:13.195056344 +1000
@@ -104,7 +104,7 @@
 
 # Camera
 PRODUCT_PACKAGES += \
-    libop2_cam
+    libop2_cam \
     Snap
 
 # Charger
@@ -284,7 +284,14 @@
     p2p_supplicant_overlay.conf \
     hostapd_default.conf \
     hostapd.accept \
-    hostapd.deny
+    hostapd.deny \
+	GmsCore \
+    FakeStore \
+	UnifiedNlp \
+	LocalGSM \
+	IchnaeaLnp \
+	ViPER4Android \
+	F-Droid 
 
 # Inherit from oppo-common
 $(call inherit-product, device/oppo/common/common.mk)
