--- device/oneplus/oneplus2/device.mk	2016-09-06 16:36:37.489489134 +1000
+++ device/oneplus/oneplus2/device.mk.new	2016-09-06 16:37:00.791378336 +1000
@@ -284,7 +284,13 @@
     p2p_supplicant_overlay.conf \
     hostapd_default.conf \
     hostapd.accept \
-    hostapd.deny 
+    hostapd.deny \
+	GmsCore \
+	FakeStore \
+	LocalGSM \
+	unifiedNlpBackend \
+	FDroid \
+	YahooWeatherProvider
 
 # Inherit from oppo-common
 $(call inherit-product, device/oppo/common/common.mk)
