--- device/oneplus/oneplus2/device.mk.orig	2016-09-06 12:44:59.458268223 +1000
+++ device/oneplus/oneplus2/device.mk	2016-09-06 12:43:03.055774141 +1000
@@ -284,7 +284,13 @@
     p2p_supplicant_overlay.conf \
     hostapd_default.conf \
     hostapd.accept \
-    hostapd.deny 
+    hostapd.deny \
+	GmsCore \
+	FakeStore \
+	LocalGSM \
+	IchnaeaLnp \
+	FDroid \
+	YahooWeatherProvider
 
 # Inherit from oppo-common
 $(call inherit-product, device/oppo/common/common.mk)
