--- device/oneplus/oneplus2/device.mk.bak	2016-09-05 13:18:46.553343826 +1000
+++ device/oneplus/oneplus2/device.mk	2016-09-05 13:19:37.814111348 +1000
@@ -284,7 +284,13 @@
     p2p_supplicant_overlay.conf \
     hostapd_default.conf \
     hostapd.accept \
-    hostapd.deny
+    hostapd.deny \
+	GmsCore \
+	FakeStore \
+	UnifiedNlp \
+	LocalGSM \
+	IchnaeaLnp \
+	YahooWeatherProvider
 
 # Inherit from oppo-common
 $(call inherit-product, device/oppo/common/common.mk)
