--- device/oneplus/oneplus2/device.mk	2016-10-19 11:56:56.955483608 +1100
+++ device/oneplus/oneplus2/device.mk.new	2016-10-19 11:57:31.215606728 +1100
@@ -271,7 +271,10 @@
     hostapd \
     libwpa_client \
     wpa_supplicant \
-    wpa_supplicant.conf
+    wpa_supplicant.conf \
+    GmsCore \
+    GsfProxy \
+    FakeStore \
+    F-Droid
 
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/wifi/hostapd.accept:system/etc/hostapd/hostapd.accept \
