--- vendor/aokp/configs/common_full.mk	2016-08-24 15:56:26.490625484 +1000
+++ vendor/aokp/configs/common_full.mk  2016-08-24 16:41:50.013250280 +1000
@@ -13,14 +13,8 @@
 # Optional AOKP packages
 PRODUCT_PACKAGES += \
     CMSettingsProvider \
-    Galaxy4 \
-    HoloSpiralWallpaper \
-    LiveWallpapers \
     LiveWallpapersPicker \
     MagicSmokeWallpapers \
-    NoiseField \
-    PhaseBeam \
-    PhotoPhase \
     PhotoTable \
     Screencast \
     SoundRecorder \
