--- device/oneplus/oneplus5/omni_oneplus5.mk	2017-10-10 16:03:31.307330437 +1100
+++ device/oneplus/oneplus5/omni_oneplus5.mk	2017-10-10 16:01:50.179330437 +1100
@@ -30,7 +30,7 @@
 $(call inherit-product, $(SRC_TARGET_DIR)/product/aosp_base_telephony.mk)
 
 # must be before including omni part
-TARGET_BOOTANIMATION_SIZE := 1080x608
+TARGET_BOOTANIMATION_SIZE := 1080p
 
 # Inherit from our custom product configuration
 $(call inherit-product, vendor/omni/config/common.mk)
