--- build/target/product/core.mk	2016-10-18 18:22:43.872121806 +1100
+++ build/target/product/core.mk.new	2016-10-18 18:37:43.685050644 +1100
@@ -25,7 +25,6 @@
     BlockedNumberProvider \
     BookmarkProvider \
     Browser \
-    Calendar \
     CalendarProvider \
     CaptivePortalLogin \
     CertInstaller \
@@ -33,7 +32,6 @@
     DeskClock \
     DocumentsUI \
     DownloadProviderUi \
-    Email \
     ExactCalculator \
     Exchange2 \
     ExternalStorageProvider \
