--- device/oneplus/onyx/device.mk.orig	2017-01-03 18:05:56.529526260 +1100
+++ device/oneplus/onyx/device.mk	2017-01-03 18:06:59.127526260 +1100
@@ -227,7 +227,20 @@
     p2p_supplicant_overlay.conf \
     hostapd_default.conf \
     hostapd.accept \
-    hostapd.deny
+    hostapd.deny \
+    F-Droid \
+    GmsCore \
+    GsfProxy \
+    FakeStore \
+    IchnaeaNlpBackend \
+    unifiedNlp \
+    GSMBackend \
+    NominatimNlpBackend \
+    OpenWeatherMapProvider \
+    WundergroundWeatherProvider \
+    YahooWeatherProvider \
+    Calendar \
+    afhDownloader
 
 PRODUCT_PACKAGES += \
     wcnss_service
