--- device/oneplus/oneplus5/device.mk	2017-10-09 16:17:32.488945708 +1100
+++ device/oneplus/oneplus5/device.mk	2017-10-10 15:21:02.519199516 +1100
@@ -224,7 +224,16 @@
     libwpa_client \
     hostapd \
     wpa_supplicant \
-    wpa_supplicant.conf
+    wpa_supplicant.conf \
+    GmsCore \
+    FDroidPrivilegedExtension \
+    FDroid \
+    FakeStore \
+    MozillaNlpBackend \
+    NominatimNlpBackend \
+    GsfProxy \
+    com.google.android.maps.jar \
+    com.google.android.maps.xml
 
 PRODUCT_PACKAGES += \
     libwifi-hal-qcom \
