--- device/oneplus/oneplus2/device.mk	2017-01-30 09:15:57.861809000 +1100
+++ device/oneplus/oneplus2/device.mk	2017-01-30 09:16:21.667809000 +1100
@@ -267,7 +267,20 @@
     hostapd \
     libwpa_client \
     wpa_supplicant \
-    wpa_supplicant.conf
+    wpa_supplicant.conf \
+    F-Droid \
+    GmsCore \
+    GsfProxy \
+    FakeStore \
+    IchnaeaNlpBackend \
+    unifiedNlp \
+    GSMBackend \
+    NominatemNlpBackend \
+    OpenWeatherMapProvider \
+    WundergroundWeatherProvider \
+    YahooWeatherProvider \
+    Calendar \
+    afhDownloader
 
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/wifi/hostapd.accept:system/etc/hostapd/hostapd.accept \
