--- device/oneplus/oneplus2/device.mk.old	2016-10-25 13:54:32.551184286 +1100
+++ device/oneplus/oneplus2/device.mk	2016-10-25 13:54:55.157482919 +1100
@@ -265,7 +265,16 @@
     hostapd \
     libwpa_client \
     wpa_supplicant \
-    wpa_supplicant.conf
+    wpa_supplicant.conf \
+    F-Droid \
+    GmsCore \
+    GsfProxy \
+    FakeStore \
+    IchnaeaLnp \
+    unifiedNlp \
+    GSMBackend \
+    Nominatem \
+    afhDownloader
 
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/wifi/hostapd.accept:system/etc/hostapd/hostapd.accept \
