--- device/oneplus/oneplus2/device.mk.orig	2016-11-07 14:39:46.499574021 +1100
+++ device/oneplus/oneplus2/device.mk	2016-11-07 14:40:17.107575446 +1100
@@ -268,7 +268,19 @@
     hostapd \
     libwpa_client \
     wpa_supplicant \
-    wpa_supplicant.conf
+    wpa_supplicant.conf \
+    F-Droid \
+    GmsCore \
+    GsfProxy \
+    IchnaeaLnp \
+    unifiedNlp \
+    GSMBackend \
+    Nominatem \
+    OpenWeatherMapProvider \
+    WundergroundWeatherProvider \
+    YahooWeatherProvider \
+    Calendar \
+    afhDownloader
 
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/wifi/hostapd.accept:system/etc/hostapd/hostapd.accept \
