--- device/oneplus/oneplus2/device.mk.orig	2016-10-30 00:40:39.845506023 +1100
+++ device/oneplus/oneplus2/device.mk	2016-10-30 00:42:57.829043021 +1100
@@ -267,7 +267,19 @@
     hostapd \
     libwpa_client \
     wpa_supplicant \
-    wpa_supplicant.conf
+    wpa_supplicant.conf \
+    F-Droid \
+    GmsCore \
+    GsfProxy \
+    FakeStore \
+    IchnaeaLnp \
+    unifiedNlp \
+    GSMBackend \
+    Nominatem \
+    OpenWeatherMapProvider \
+    WundergroundWeatherProvider \
+    YahooWeatherProvider \
+    afhDownloader
 
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/wifi/hostapd.accept:system/etc/hostapd/hostapd.accept \
