--- packages/apps/NominatemNlpBackend/Android.mk	2017-01-30 15:10:33.998000000 +1100
+++ packages/apps/NominatemNlpBackend/Android.mk	2017-01-30 15:10:42.191000000 +1100
@@ -9,7 +9,7 @@
 nominatim_root  := $(LOCAL_PATH)
 nominatim_out   := $(OUT_DIR)/target/common/obj/APPS/$(LOCAL_MODULE)_intermediates
 nominatim_build := $(nominatim_root)/build
-nominatim_apk   := build/outputs/apk/NominatimGeocoderBackend-release-unsigned.apk
+nominatim_apk   := build/outputs/apk/NominatemNlpBackend-release-unsigned.apk
 
 $(nominatim_root)/$(nominatim_apk):
 	rm -Rf $(nominatim_build)
