--- frameworks/base/core/java/android/hardware/camera2/legacy/LegacyMetadataMapper.java	2017-10-09 14:47:58.922945708 +1100
+++ LegacyMetadataMapper.java	2017-10-26 15:59:48.695711730 +1100
@@ -16,6 +16,7 @@
 
 package android.hardware.camera2.legacy;
 
+import android.app.ActivityThread;
 import android.graphics.ImageFormat;
 import android.graphics.PixelFormat;
 import android.graphics.Rect;
@@ -159,6 +160,16 @@
             Log.v(TAG, "--------------------------------------------------- (end)");
         }
 
+	String packageName = ActivityThread.currentOpPackageName();
+	if (packageName != null) {
+	    String focusMode = params.getFocusMode();
+	    if ((focusMode != null) && (focusMode.equals("fixed")) && (packageName.equals("com.oneplus.camera"))) {
+		return new CameraCharacteristics(m);
+	    }
+	}
+        Log.v("LegacyMetadataMapper", "front camera flash not available for non-opcamera");
+        m.set(CameraCharacteristics.FLASH_INFO_AVAILABLE, Boolean.valueOf(false));
+
         return new CameraCharacteristics(m);
     }
 
