diff -uNar packages/apps/LocalGSMBackend/Android.mk packages/apps/LocalGSMBackend/Android.mk
--- packages/apps/LocalGSMBackend/Android.mk	1970-01-01 10:00:00.000000000 +1000
+++ packages/apps/LocalGSMBackend/Android.mk	2016-09-05 13:13:25.174806816 +1000
@@ -0,0 +1,26 @@
+LOCAL_PATH:= $(call my-dir)
+
+include $(CLEAR_VARS)
+
+LOCAL_MODULE := GSMBackend
+LOCAL_MODULE_TAGS := optional
+LOCAL_PACKAGE_NAME := GSMBackend
+
+gsmback_root  := $(LOCAL_PATH)
+gsmback_dir   := app
+gsmback_out   := $(OUT_DIR)/target/common/obj/APPS/$(LOCAL_MODULE)_intermediates
+gsmback_build := $(gsmback_root)/$(gsmback_dir)/build
+gsmback_apk   := build/outputs/apk/$(gsmback_dir)-release-unsigned.apk
+
+$(gsmback_root)/$(gsmback_dir)/$(gsmback_apk):
+	rm -Rf $(gsmback_build)
+	mkdir -p $(gsmback_out)
+	ln -sf $(gsmback_out) $(gsmback_build)
+	cd $(gsmback_root)/$(gsmback_dir) && JAVA_TOOL_OPTIONS="$(JAVA_TOOL_OPTIONS) -Dfile.encoding=UTF8" ../gradlew assembleRelease
+
+LOCAL_CERTIFICATE := platform
+LOCAL_SRC_FILES := $(gsmback_dir)/$(gsmback_apk)
+LOCAL_MODULE_CLASS := APPS
+LOCAL_MODULE_SUFFIX := $(COMMON_ANDROID_PACKAGE_SUFFIX)
+
+include $(BUILD_PREBUILT)
