--- packages/apps/LocalGSMBackend/app/build.gradle	2016-12-18 22:21:59.603000000 +1100
+++ packages/apps/LocalGSMBackend/app/build.gradle	2016-12-18 22:22:07.432000000 +1100
@@ -39,6 +39,13 @@
         exclude 'META-INF/LICENSE.txt'
         exclude 'META-INF/NOTICE.txt'
     }
+      lintOptions {
+          checkReleaseBuilds false
+          // Or, if you prefer, you can continue to check for errors in release builds,
+          // but continue the build even when errors are found:
+          abortOnError false
+      }
+
 }
 
 dependencies {
