diff --git a/adbd.te b/adbd.te
index 9dc41fd..83a271b 100644
--- system/sepolicy/adbd.te
+++ system/sepolicy/adbd.te
@@ -111,3 +111,5 @@ allow adbd mnt_user_file:lnk_file r_file_perms;
 # accesses to the underlying FS.
 allow adbd media_rw_data_file:dir create_dir_perms;
 allow adbd media_rw_data_file:file create_file_perms;
+
+r_dir_file(adbd, apk_data_file)
diff --git a/app.te b/app.te
index f2adf37..e9dd7b3 100644
--- system/sepolicy/app.te
+++ system/sepolicy/app.te
@@ -438,6 +438,21 @@ neverallow appdomain {
   tmpfs
 }:lnk_file no_w_file_perms;
 
+# Blacklist app domains not allowed to execute from /data
+neverallow {
+  bluetooth
+  isolated_app
+  nfc
+  radio
+  shared_relro
+  system_app
+} {
+  data_file_type
+  -dalvikcache_data_file
+  -system_data_file # shared libs in apks
+  -apk_data_file
+}:file no_x_file_perms;
+
 # Foreign dex profiles are just markers. Prevent apps to do anything but touch them.
 neverallow appdomain user_profile_foreign_dex_data_file:file rw_file_perms;
 neverallow appdomain user_profile_foreign_dex_data_file:dir { open getattr read ioctl remove_name };
diff --git a/attributes b/attributes
index 1160a95..a846c34 100644
--- system/sepolicy/attributes
+++ system/sepolicy/attributes
@@ -108,3 +108,8 @@ attribute binderservicedomain;
 # requires are specific to the implementation provided in each device, but
 # common daemons need to be aware of those when calling into the HAL.
 attribute boot_control_hal;
+
+# update_engine related domains that need to apply an update and run
+# postinstall. This includes the background daemon and the sideload tool from
+# recovery for A/B devices.
+attribute update_engine_common;
diff --git a/bootanim.te b/bootanim.te
index 91a50d5..c3091ab 100644
--- system/sepolicy/bootanim.te
+++ system/sepolicy/bootanim.te
@@ -6,6 +6,7 @@ init_daemon_domain(bootanim)
 
 binder_use(bootanim)
 binder_call(bootanim, surfaceflinger)
+binder_call(bootanim, audioserver)
 
 allow bootanim gpu_device:chr_file rw_file_perms;
 
@@ -16,7 +17,9 @@ allow bootanim oemfs:file r_file_perms;
 allow bootanim audio_device:dir r_dir_perms;
 allow bootanim audio_device:chr_file rw_file_perms;
 
+allow bootanim audioserver_service:service_manager find;
 allow bootanim surfaceflinger_service:service_manager find;
+allow bootanim audioserver_service:service_manager find;
 
 # Allow access to ion memory allocation device
 allow bootanim ion_device:chr_file rw_file_perms;
diff --git a/cppreopts.te b/cppreopts.te
new file mode 100644
index 0000000..66df7ee
--- /dev/null
+++ system/sepolicy/cppreopts.te
@@ -0,0 +1,28 @@
+# cppreopts
+#
+# This command copies preopted files from the system_b partition to the data
+# partition. This domain ensures that we are only copying into specific
+# directories.
+
+type cppreopts, domain, mlstrustedsubject;
+type cppreopts_exec, exec_type, file_type;
+
+# Technically not a daemon but we do want the transition from init domain to
+# cppreopts to occur.
+init_daemon_domain(cppreopts)
+
+domain_auto_trans(cppreopts, preopt2cachename_exec, preopt2cachename);
+
+# Allow cppreopts copy files into the dalvik-cache
+allow cppreopts dalvikcache_data_file:dir { add_name remove_name search write };
+allow cppreopts dalvikcache_data_file:file { create getattr open read rename write };
+
+# Allow cppreopts to execute itself using #!/system/bin/sh
+allow cppreopts shell_exec:file rx_file_perms;
+
+# Allow us to run find on /postinstall
+allow cppreopts system_file:dir { open read };
+
+# Allow running the cp command using cppreopts permissions. Needed so we can
+# write into dalvik-cache
+allow cppreopts toolbox_exec:file rx_file_perms;
diff --git a/dex2oat.te b/dex2oat.te
index 48daac3..fdf5536 100644
--- system/sepolicy/dex2oat.te
+++ system/sepolicy/dex2oat.te
@@ -24,7 +24,7 @@ allow dex2oat user_profile_data_file:file { getattr read lock };
 # Allow dex2oat to use file descriptors from otapreopt.
 allow dex2oat postinstall_dexopt:fd use;
 
-allow dex2oat postinstall_file:dir getattr;
+allow dex2oat postinstall_file:dir { getattr search };
 
 # Allow dex2oat access to files in /data/ota.
 allow dex2oat ota_data_file:dir ra_dir_perms;
diff --git a/domain.te b/domain.te
index 1ac33f1..45569de 100644
--- system/sepolicy/domain.te
+++ system/sepolicy/domain.te
@@ -265,7 +265,7 @@ neverallow { domain -init -ueventd } device:chr_file { open read write };
 # Limit what domains can mount filesystems or change their mount flags.
 # sdcard_type / vfat is exempt as a larger set of domains need
 # this capability, including device-specific domains.
-neverallow { domain -kernel -init -recovery -vold -zygote -update_engine } { fs_type -sdcard_type }:filesystem { mount remount relabelfrom relabelto };
+neverallow { domain -kernel -init -recovery -vold -zygote -update_engine -otapreopt_chroot } { fs_type -sdcard_type }:filesystem { mount remount relabelfrom relabelto };
 
 #
 # Assert that, to the extent possible, we're not loading executable content from
@@ -292,9 +292,7 @@ neverallow domain { cache_file cache_backup_file cache_private_backup_file cache
 # Protect most domains from executing arbitrary content from /data.
 neverallow {
   domain
-  -untrusted_app
-  -priv_app
-  -shell
+  -appdomain
 } {
   data_file_type
   -dalvikcache_data_file
@@ -378,7 +376,9 @@ neverallow {
   -zygote
   -installd
   -postinstall_dexopt
+  -cppreopts
   -dex2oat
+  -otapreopt_slot
 } dalvikcache_data_file:file no_w_file_perms;
 
 neverallow {
@@ -386,8 +386,10 @@ neverallow {
   -init
   -installd
   -postinstall_dexopt
+  -cppreopts
   -dex2oat
   -zygote
+  -otapreopt_slot
 } dalvikcache_data_file:dir no_w_dir_perms;
 
 # Only system_server should be able to send commands via the zygote socket
@@ -485,6 +487,11 @@ neverallow {
   -zygote
 } shell:process { transition dyntransition };
 
+# Only domains spawned from zygote and runas may have the appdomain attribute.
+neverallow { domain -runas -zygote } {
+  appdomain -shell userdebug_or_eng(`-su') -bluetooth
+}:process { transition dyntransition };
+
 # Minimize read access to shell- or app-writable symlinks.
 # This is to prevent malicious symlink attacks.
 neverallow {
diff --git a/drmserver.te b/drmserver.te
index 9a9cfc0..9130e0b 100644
--- system/sepolicy/drmserver.te
+++ system/sepolicy/drmserver.te
@@ -41,6 +41,7 @@ r_dir_file(drmserver, media_rw_data_file)
 # Read resources from open apk files passed over Binder.
 allow drmserver apk_data_file:file { read getattr };
 allow drmserver asec_apk_file:file { read getattr };
+allow drmserver ringtone_file:file { read getattr };
 
 # Read /data/data/com.android.providers.telephony files passed over Binder.
 allow drmserver radio_data_file:file { read getattr };
diff --git a/dumpstate.te b/dumpstate.te
index 3187555..115bb09 100644
--- system/sepolicy/dumpstate.te
+++ system/sepolicy/dumpstate.te
@@ -5,6 +5,7 @@ type dumpstate_exec, exec_type, file_type;
 init_daemon_domain(dumpstate)
 net_domain(dumpstate)
 binder_use(dumpstate)
+wakelock_use(dumpstate)
 
 # Allow setting process priority, protect from OOM killer, and dropping
 # privileges by switching UID / GID
@@ -136,6 +137,9 @@ control_logd(dumpstate)
 allow dumpstate net_data_file:dir search;
 allow dumpstate net_data_file:file r_file_perms;
 
+# List sockets via ss.
+allow dumpstate self:netlink_tcpdiag_socket { create_socket_perms nlmsg_read };
+
 # Access /data/tombstones.
 allow dumpstate tombstone_data_file:dir r_dir_perms;
 allow dumpstate tombstone_data_file:file r_file_perms;
@@ -148,6 +152,14 @@ allow dumpstate cache_recovery_file:file r_file_perms;
 allow dumpstate recovery_data_file:dir r_dir_perms;
 allow dumpstate recovery_data_file:file r_file_perms;
 
+# Access /data/misc/profiles/{cur,ref}/
+userdebug_or_eng(`
+  allow dumpstate user_profile_data_file:dir r_dir_perms;
+  allow dumpstate user_profile_data_file:file r_file_perms;
+  allow dumpstate user_profile_foreign_dex_data_file:dir r_dir_perms;
+  allow dumpstate user_profile_foreign_dex_data_file:file r_file_perms;
+')
+
 # Access /data/misc/logd
 userdebug_or_eng(`
   allow dumpstate misc_logd_file:dir r_dir_perms;
diff --git a/file.te b/file.te
index f0e984d..84af4a7 100644
--- system/sepolicy/file.te
+++ system/sepolicy/file.te
@@ -92,6 +92,8 @@ type apk_private_tmp_file, file_type, data_file_type, mlstrustedobject;
 type dalvikcache_data_file, file_type, data_file_type;
 # /data/ota
 type ota_data_file, file_type, data_file_type;
+# /data/ota_package
+type ota_package_file, file_type, data_file_type, mlstrustedobject;
 # /data/misc/profiles
 type user_profile_data_file, file_type, data_file_type, mlstrustedobject;
 type user_profile_foreign_dex_data_file, file_type, data_file_type, mlstrustedobject;
diff --git a/file_contexts b/file_contexts
index 0192ea3..3d91e2b 100644
--- system/sepolicy/file_contexts
+++ system/sepolicy/file_contexts
@@ -175,6 +175,7 @@
 /system/bin/mdnsd	u:object_r:mdnsd_exec:s0
 /system/bin/installd	u:object_r:installd_exec:s0
 /system/bin/otapreopt_chroot   u:object_r:otapreopt_chroot_exec:s0
+/system/bin/otapreopt_slot   u:object_r:otapreopt_slot_exec:s0
 /system/bin/keystore	u:object_r:keystore_exec:s0
 /system/bin/fingerprintd u:object_r:fingerprintd_exec:s0
 /system/bin/gatekeeperd u:object_r:gatekeeperd_exec:s0
@@ -203,6 +204,8 @@
 /system/bin/update_verifier u:object_r:update_verifier_exec:s0
 /system/bin/logwrapper  u:object_r:system_file:s0
 /system/bin/vdc         u:object_r:vdc_exec:s0
+/system/bin/cppreopts.sh   u:object_r:cppreopts_exec:s0
+/system/bin/preopt2cachename u:object_r:preopt2cachename_exec:s0
 /system/bin/install-recovery.sh u:object_r:install_recovery_exec:s0
 /system/bin/dex2oat     u:object_r:dex2oat_exec:s0
 # patchoat executable has (essentially) the same requirements as dex2oat.
@@ -246,6 +249,7 @@
 /data/resource-cache(/.*)? u:object_r:resourcecache_data_file:s0
 /data/dalvik-cache(/.*)? u:object_r:dalvikcache_data_file:s0
 /data/ota(/.*)? u:object_r:ota_data_file:s0
+/data/ota_package(/.*)? u:object_r:ota_package_file:s0
 /data/adb(/.*)?		u:object_r:adb_data_file:s0
 /data/anr(/.*)?		u:object_r:anr_data_file:s0
 /data/app(/.*)?                       u:object_r:apk_data_file:s0
diff --git a/init.te b/init.te
index 6197c39..9bc78d1 100644
--- system/sepolicy/init.te
+++ system/sepolicy/init.te
@@ -44,7 +44,7 @@ allow init self:capability sys_admin;
 
 # Create and mount on directories in /.
 allow init rootfs:dir create_dir_perms;
-allow init { rootfs cache_file cgroup storage_file system_data_file system_file }:dir mounton;
+allow init { rootfs cache_file cgroup storage_file system_data_file system_file postinstall_mnt_dir }:dir mounton;
 
 # Mount on /dev/usb-ffs/adb.
 allow init device:dir mounton;
diff --git a/installd.te b/installd.te
index ebd7591..ab0aadc 100644
--- system/sepolicy/installd.te
+++ system/sepolicy/installd.te
@@ -72,11 +72,6 @@ domain_auto_trans(installd, profman_exec, profman)
 # Run idmap in its own sandbox.
 domain_auto_trans(installd, idmap_exec, idmap)
 
-# Run otapreopt in its own sandbox.
-domain_auto_trans(installd, otapreopt_chroot_exec, otapreopt_chroot)
-# otapreopt_chroot will transition into postinstall_dexopt, which will spawn a child.
-allow installd postinstall_dexopt:process sigchld;
-
 # Upgrade from unlabeled userdata.
 # Just need enough to remove and/or relabel it.
 allow installd unlabeled:dir { getattr search relabelfrom rw_dir_perms rmdir };
diff --git a/ioctl_defines b/ioctl_defines
index 5b65b2d..e879b38 100644
--- system/sepolicy/ioctl_defines
+++ system/sepolicy/ioctl_defines
@@ -681,8 +681,6 @@ define(`BR_NOOP', `0x0000720c')
 define(`BR_SPAWN_LOOPER', `0x0000720d')
 define(`BR_FINISHED', `0x0000720e')
 define(`BR_FAILED_REPLY', `0x00007211')
-define(`PPPIOCDISCONN', `0x00007439')
-define(`PPPIOCXFERUNIT', `0x0000744e')
 define(`MEYEIOC_STILLCAPT', `0x000076c4')
 define(`ASHMEM_GET_SIZE', `0x00007704')
 define(`ASHMEM_GET_PROT_MASK', `0x00007706')
@@ -1166,22 +1164,6 @@ define(`IXJCTL_SC_TXG', `0x400471eb')
 define(`IXJCTL_INTERCOM_START', `0x400471fd')
 define(`IXJCTL_INTERCOM_STOP', `0x400471fe')
 define(`FAT_IOCTL_SET_ATTRIBUTES', `0x40047211')
-define(`PPPIOCATTCHAN', `0x40047438')
-define(`PPPIOCCONNECT', `0x4004743a')
-define(`PPPIOCSMRRU', `0x4004743b')
-define(`PPPIOCDETACH', `0x4004743c')
-define(`PPPIOCATTACH', `0x4004743d')
-define(`PPPIOCSDEBUG', `0x40047440')
-define(`PPPIOCSMAXCID', `0x40047451')
-define(`PPPIOCSMRU', `0x40047452')
-define(`PPPIOCSRASYNCMAP', `0x40047454')
-define(`PPPIOCSASYNCMAP', `0x40047457')
-define(`PPPIOCSFLAGS', `0x40047459')
-define(`PPPIOCBUNDLE', `0x40047481')
-define(`PPPIOCSMPFLAGS', `0x40047483')
-define(`PPPIOCSMPMTU', `0x40047484')
-define(`PPPIOCSMPMRU', `0x40047485')
-define(`PPPIOCSCOMPRESSOR', `0x40047487')
 define(`V4L2_SUBDEV_IR_RX_NOTIFY', `0x40047600')
 define(`V4L2_SUBDEV_IR_TX_NOTIFY', `0x40047601')
 define(`FS_IOC32_SETVERSION', `0x40047602')
@@ -1304,7 +1286,6 @@ define(`IXJCTL_FILTER_CADENCE', `0x400871d6')
 define(`IXJCTL_CIDCW', `0x400871d9')
 define(`IXJCTL_SET_FILTER_RAW', `0x400871dd')
 define(`IXJCTL_SIGCTL', `0x400871e9')
-define(`PPPIOCSNPMODE', `0x4008744b')
 define(`FS_IOC_SETVERSION', `0x40087602')
 define(`ASHMEM_SET_SIZE', `0x40087703')
 define(`ASHMEM_SET_PROT_MASK', `0x40087705')
@@ -1453,9 +1434,6 @@ define(`VIDEO_SET_SPU_PALETTE', `0x40106f33')
 define(`FE_SET_PROPERTY', `0x40106f52')
 define(`CA_SET_DESCR', `0x40106f86')
 define(`PPSETTIME', `0x40107096')
-define(`PPPIOCSACTIVE', `0x40107446')
-define(`PPPIOCSPASS', `0x40107447')
-define(`PPPIOCSCOMPRESS', `0x4010744d')
 define(`BTRFS_IOC_QGROUP_CREATE', `0x4010942a')
 define(`GENWQE_WRITE_REG64', `0x4010a51f')
 define(`GENWQE_WRITE_REG32', `0x4010a521')
@@ -1530,7 +1508,6 @@ define(`DRM_IOCTL_I915_GEM_PREAD', `0x4020645c')
 define(`DRM_IOCTL_I915_GEM_PWRITE', `0x4020645d')
 define(`OSD_SEND_CMD', `0x40206fa0')
 define(`RTC_PLL_SET', `0x40207012')
-define(`PPPIOCSXASYNCMAP', `0x4020744f')
 define(`BTRFS_IOC_CLONE_RANGE', `0x4020940d')
 define(`KVM_SET_MEMORY_ALIAS', `0x4020ae43')
 define(`KVM_SET_USER_MEMORY_REGION', `0x4020ae46')
@@ -1875,14 +1852,6 @@ define(`BR_ERROR', `0x80047200')
 define(`BR_ACQUIRE_RESULT', `0x80047204')
 define(`FAT_IOCTL_GET_ATTRIBUTES', `0x80047210')
 define(`FAT_IOCTL_GET_VOLUME_ID', `0x80047213')
-define(`PPPIOCGCHAN', `0x80047437')
-define(`PPPIOCGDEBUG', `0x80047441')
-define(`PPPIOCGMRU', `0x80047453')
-define(`PPPIOCGRASYNCMAP', `0x80047455')
-define(`PPPIOCGUNIT', `0x80047456')
-define(`PPPIOCGASYNCMAP', `0x80047458')
-define(`PPPIOCGFLAGS', `0x8004745a')
-define(`PPPIOCGMPFLAGS', `0x80047482')
 define(`FS_IOC32_GETVERSION', `0x80047601')
 define(`MEYEIOC_STILLJCAPT', `0x800476c5')
 define(`OSIOCGNETADDR', `0x800489e1')
@@ -2015,8 +1984,6 @@ define(`BR_INCREFS', `0x80107207')
 define(`BR_ACQUIRE', `0x80107208')
 define(`BR_RELEASE', `0x80107209')
 define(`BR_DECREFS', `0x8010720a')
-define(`PPPIOCGIDLE', `0x8010743f')
-define(`PPPIOCGIFNAME', `0x80107488')
 define(`GENWQE_READ_REG64', `0x8010a51e')
 define(`GENWQE_READ_REG32', `0x8010a520')
 define(`GENWQE_READ_REG16', `0x8010a522')
@@ -2054,7 +2021,6 @@ define(`I2OGETIOPS', `0x80206900')
 define(`AUDIO_GET_STATUS', `0x80206f0a')
 define(`VIDEO_GET_EVENT', `0x80206f1c')
 define(`RTC_PLL_GET', `0x80207011')
-define(`PPPIOCGXASYNCMAP', `0x80207450')
 define(`KVM_ARM_PREFERRED_TARGET', `0x8020aeaf')
 define(`SNDRV_HDSP_IOCTL_GET_CONFIG_INFO', `0x80244841')
 define(`SNDRV_HDSPM_IOCTL_GET_VERSION', `0x80244848')
@@ -2093,12 +2059,10 @@ define(`SNDRV_HWDEP_IOCTL_DSP_STATUS', `0x80404802')
 define(`JSIOCGAXMAP', `0x80406a32')
 define(`BR_TRANSACTION', `0x80407202')
 define(`BR_REPLY', `0x80407203')
-define(`PPPIOCGCOMPRESSORS', `0x80407486')
 define(`BTRFS_IOC_QUOTA_RESCAN_STATUS', `0x8040942d')
 define(`KVM_ASSIGN_PCI_DEVICE', `0x8040ae69')
 define(`KVM_GET_VCPU_EVENTS', `0x8040ae9f')
 define(`GET_ARRAY_INFO', `0x80480911')
-define(`PPPIOCGL2TPSTATS', `0x80487436')
 define(`BTRFS_IOC_GET_SUPPORTED_FEATURES', `0x80489439')
 define(`KVM_SET_PIT', `0x8048ae66')
 define(`GSMIOC_GETCONF', `0x804c4700')
@@ -2213,7 +2177,6 @@ define(`DRM_IOCTL_MODE_DESTROY_DUMB', `0xc00464b4')
 define(`SNDCTL_MIDI_PRETIME', `0xc0046d00')
 define(`SNDCTL_MIDI_MPUMODE', `0xc0046d01')
 define(`MGSL_IOCWAITEVENT', `0xc0046d08')
-define(`PPPIOCNEWUNIT', `0xc004743e')
 define(`TOSH_SMM', `0xc0047490')
 define(`MEYEIOC_SYNC', `0xc00476c3')
 define(`AUTOFS_IOC_SETTIMEOUT32', `0xc0049364')
@@ -2273,7 +2236,6 @@ define(`PHONE_QUERY_CODEC', `0xc00871a7')
 define(`MIC_VIRTIO_ADD_DEVICE', `0xc0087301')
 define(`MIC_VIRTIO_COPY_DESC', `0xc0087302')
 define(`MIC_VIRTIO_CONFIG_CHANGE', `0xc0087305')
-define(`PPPIOCGNPMODE', `0xc008744c')
 define(`AUTOFS_IOC_SETTIMEOUT', `0xc0089364')
 define(`KVM_GET_SUPPORTED_CPUID', `0xc008ae05')
 define(`KVM_GET_EMULATED_CPUID', `0xc008ae09')
@@ -2616,7 +2578,6 @@ define(`VIDIOC_G_DV_TIMINGS', `0xc0845658')
 define(`VIDIOC_SUBDEV_G_DV_TIMINGS', `0xc0845658')
 define(`SNDRV_PCM_IOCTL_SW_PARAMS', `0xc0884113')
 define(`SNDRV_PCM_IOCTL_SYNC_PTR', `0xc0884123')
-define(`PPPIOCGCALLINFO', `0xc0887480')
 define(`SNDCTL_SYNTH_INFO', `0xc08c5102')
 define(`SNDCTL_SYNTH_ID', `0xc08c5114')
 define(`SNDRV_SEQ_IOCTL_CREATE_QUEUE', `0xc08c5332')
@@ -2692,3 +2653,42 @@ define(`HIDIOCGUSAGES', `0xd01c4813')
 define(`SNDRV_COMPRESS_GET_CODEC_CAPS', `0xeb884311')
 define(`WAN_IOC_ADD_FLT_RULE', `0x00006900')
 define(`WAN_IOC_ADD_FLT_INDEX', `0x00006902')
+define(`PPPIOCGL2TPSTATS',   `0x7436')
+define(`PPPIOCGCHAN',        `0x7437')
+define(`PPPIOCATTCHAN',      `0x7438')
+define(`PPPIOCDISCONN',      `0x7439')
+define(`PPPIOCCONNECT',      `0x743a')
+define(`PPPIOCSMRRU',        `0x743b')
+define(`PPPIOCDETACH',       `0x743c')
+define(`PPPIOCATTACH',       `0x743d')
+define(`PPPIOCNEWUNIT',      `0x743e')
+define(`PPPIOCGIDLE',        `0x743f')
+define(`PPPIOCSDEBUG',       `0x7440')
+define(`PPPIOCGDEBUG',       `0x7441')
+define(`PPPIOCSACTIVE',      `0x7446')
+define(`PPPIOCSPASS',        `0x7447')
+define(`PPPIOCSNPMODE',      `0x744b')
+define(`PPPIOCGNPMODE',      `0x744c')
+define(`PPPIOCSCOMPRESS',    `0x744d')
+define(`PPPIOCXFERUNIT',     `0x744e')
+define(`PPPIOCSXASYNCMAP',   `0x744f')
+define(`PPPIOCGXASYNCMAP',   `0x7450')
+define(`PPPIOCSMAXCID',      `0x7451')
+define(`PPPIOCSMRU',         `0x7452')
+define(`PPPIOCGMRU',         `0x7453')
+define(`PPPIOCSRASYNCMAP',   `0x7454')
+define(`PPPIOCGRASYNCMAP',   `0x7455')
+define(`PPPIOCGUNIT',        `0x7456')
+define(`PPPIOCSASYNCMAP',    `0x7457')
+define(`PPPIOCGASYNCMAP',    `0x7458')
+define(`PPPIOCSFLAGS',       `0x7459')
+define(`PPPIOCGFLAGS',       `0x745a')
+define(`PPPIOCGCALLINFO',    `0x7480')
+define(`PPPIOCBUNDLE',       `0x7481')
+define(`PPPIOCGMPFLAGS',     `0x7482')
+define(`PPPIOCSMPFLAGS',     `0x7483')
+define(`PPPIOCSMPMTU',       `0x7484')
+define(`PPPIOCSMPMRU',       `0x7485')
+define(`PPPIOCGCOMPRESSORS', `0x7486')
+define(`PPPIOCSCOMPRESSOR',  `0x7487')
+define(`PPPIOCGIFNAME',      `0x7488')
diff --git a/ioctl_macros b/ioctl_macros
index 858bd78..f3840b8 100644
--- system/sepolicy/ioctl_macros
+++ system/sepolicy/ioctl_macros
@@ -46,3 +46,17 @@ define(`unpriv_unix_sock_ioctls', `{TIOCOUTQ FIOCLEX TCGETS TIOCGWINSZ TIOCSWINS
 
 # commonly used TTY ioctls
 define(`unpriv_tty_ioctls', `{ TIOCOUTQ FIOCLEX }')
+
+# point to point ioctls
+define(`ppp_ioctls', `{
+PPPIOCGL2TPSTATS PPPIOCGCHAN PPPIOCATTCHAN PPPIOCDISCONN
+PPPIOCCONNECT PPPIOCSMRRU PPPIOCDETACH PPPIOCATTACH
+PPPIOCNEWUNIT PPPIOCGIDLE PPPIOCSDEBUG PPPIOCGDEBUG
+PPPIOCSACTIVE PPPIOCSPASS PPPIOCSNPMODE PPPIOCGNPMODE
+PPPIOCSCOMPRESS PPPIOCXFERUNIT PPPIOCSXASYNCMAP
+PPPIOCGXASYNCMAP PPPIOCSMAXCID PPPIOCSMRU PPPIOCGMRU
+PPPIOCSRASYNCMAP PPPIOCGRASYNCMAP PPPIOCGUNIT PPPIOCSASYNCMAP
+PPPIOCGASYNCMAP PPPIOCSFLAGS PPPIOCGFLAGS PPPIOCGCALLINFO
+PPPIOCBUNDLE PPPIOCGMPFLAGS PPPIOCSMPFLAGS PPPIOCSMPMTU
+PPPIOCSMPMRU PPPIOCGCOMPRESSORS PPPIOCSCOMPRESSOR PPPIOCGIFNAME
+}')
diff --git a/logd.te b/logd.te
index 97bbd8b..bc97a37 100644
--- system/sepolicy/logd.te
+++ system/sepolicy/logd.te
@@ -28,6 +28,7 @@ set_prop(logd, powerctl_prop)
 
 # Access device logging gating property
 get_prop(logd, device_logging_prop)
+userdebug_or_eng(`get_prop(logd, logpersistd_logging_prop)')
 
 r_dir_file(logd, domain)
 
diff --git a/mediadrmserver.te b/mediadrmserver.te
index cfa4b28..d9368ad 100644
--- system/sepolicy/mediadrmserver.te
+++ system/sepolicy/mediadrmserver.te
@@ -49,6 +49,7 @@ allow mediadrmserver tee:unix_stream_socket connectto;
 allow mediadrmserver mediadrmserver_service:service_manager { add find };
 allow mediadrmserver mediaserver_service:service_manager { add find };
 allow mediadrmserver processinfo_service:service_manager find;
+allow mediadrmserver surfaceflinger_service:service_manager find;
 
 # only allow unprivileged socket ioctl commands
 allowxperm mediadrmserver self:{ rawip_socket tcp_socket udp_socket }
diff --git a/nfc.te b/nfc.te
index 5b7f4b9..fc7e167 100644
--- system/sepolicy/nfc.te
+++ system/sepolicy/nfc.te
@@ -21,10 +21,11 @@ allow nfc sysfs:file write;
 allow nfc sysfs_usb:file write;
 
 # SoundPool loading and playback
-allow nfc mediaserver_service:service_manager find;
 allow nfc audioserver_service:service_manager find;
-allow nfc mediaextractor_service:service_manager find;
+allow nfc drmserver_service:service_manager find;
 allow nfc mediacodec_service:service_manager find;
+allow nfc mediaextractor_service:service_manager find;
+allow nfc mediaserver_service:service_manager find;
 
 allow nfc nfc_service:service_manager { add find };
 allow nfc radio_service:service_manager find;
diff --git a/otapreopt_chroot.te b/otapreopt_chroot.te
index b3f8807..1c5f2ee 100644
--- system/sepolicy/otapreopt_chroot.te
+++ system/sepolicy/otapreopt_chroot.te
@@ -7,8 +7,17 @@ type otapreopt_chroot_exec, exec_type, file_type;
 allow otapreopt_chroot postinstall_file:dir { search mounton };
 allow otapreopt_chroot self:capability { sys_admin sys_chroot };
 
+# This is required to mount /vendor.
+allow otapreopt_chroot block_device:dir search;
+allow otapreopt_chroot labeledfs:filesystem mount;
+# Mounting /vendor can have this side-effect. Ignore denial.
+dontaudit otapreopt_chroot kernel:process setsched;
+
 # Allow to transition to postinstall_ota, to run otapreopt in its own sandbox.
 domain_auto_trans(otapreopt_chroot, postinstall_file, postinstall_dexopt)
 
-# Allow otapreopt to use file descriptors from installd.
-allow otapreopt_chroot installd:fd use;
+# Allow otapreopt to use file descriptors from update-engine. It will
+# close them immediately.
+allow otapreopt_chroot postinstall:fd use;
+allow otapreopt_chroot update_engine:fd use;
+allow otapreopt_chroot update_engine:fifo_file write;
diff --git a/otapreopt_slot.te b/otapreopt_slot.te
new file mode 100644
index 0000000..2f4da0a
--- /dev/null
+++ system/sepolicy/otapreopt_slot.te
@@ -0,0 +1,28 @@
+# otapreopt_slot
+#
+# This command set moves the artifact corresponding to the current slot
+# from /data/ota to /data/dalvik-cache.
+
+type otapreopt_slot, domain, mlstrustedsubject;
+type otapreopt_slot_exec, exec_type, file_type;
+
+# Technically not a daemon but we do want the transition from init domain to
+# cppreopts to occur.
+init_daemon_domain(otapreopt_slot)
+
+# The otapreopt_slot renames the OTA dalvik-cache to the regular dalvik-cache, and cleans up
+# the directory afterwards. For logging of aggregate size, we need getattr.
+allow otapreopt_slot ota_data_file:dir { rw_dir_perms rename reparent rmdir };
+allow otapreopt_slot ota_data_file:file { getattr };
+
+# Delete old content of the dalvik-cache.
+allow otapreopt_slot dalvikcache_data_file:dir { add_name getattr open read remove_name rmdir search write };
+allow otapreopt_slot dalvikcache_data_file:file { getattr unlink };
+allow otapreopt_slot dalvikcache_data_file:lnk_file { getattr read unlink };
+
+# Allow cppreopts to execute itself using #!/system/bin/sh
+allow otapreopt_slot shell_exec:file rx_file_perms;
+
+# Allow running the mv and rm/rmdir commands using otapreopt_slot  permissions.
+# Needed so we can move artifacts into /data/dalvik-cache/dalvik-cache.
+allow otapreopt_slot toolbox_exec:file rx_file_perms;
diff --git a/platform_app.te b/platform_app.te
index 0d3bdba..d4a27ad 100644
--- system/sepolicy/platform_app.te
+++ system/sepolicy/platform_app.te
@@ -45,6 +45,7 @@ allow platform_app drmserver_service:service_manager find;
 allow platform_app mediaserver_service:service_manager find;
 allow platform_app mediaextractor_service:service_manager find;
 allow platform_app mediacodec_service:service_manager find;
+allow platform_app mediadrmserver_service:service_manager find;
 allow platform_app persistent_data_block_service:service_manager find;
 allow platform_app radio_service:service_manager find;
 allow platform_app surfaceflinger_service:service_manager find;
diff --git a/postinstall.te b/postinstall.te
index 5c261ef..0f6bb74 100644
--- system/sepolicy/postinstall.te
+++ system/sepolicy/postinstall.te
@@ -5,8 +5,8 @@ type postinstall, domain;
 
 # Allow postinstall to write to its stdout/stderr when redirected via pipes to
 # update_engine.
-allow postinstall update_engine:fd use;
-allow postinstall update_engine:fifo_file rw_file_perms;
+allow postinstall update_engine_common:fd use;
+allow postinstall update_engine_common:fifo_file rw_file_perms;
 
 # Allow postinstall to read and execute directories and files in the same
 # mounted location.
@@ -19,10 +19,6 @@ allow postinstall shell_exec:file rx_file_perms;
 allow postinstall system_file:file rx_file_perms;
 allow postinstall toolbox_exec:file rx_file_perms;
 
-# No domain other than update_engine should transition to postinstall, as it is
-# only meant to run during the update.
-neverallow { domain -update_engine } postinstall:process { transition dyntransition };
-
 #
 # For OTA dexopt.
 #
@@ -32,4 +28,11 @@ binder_use(postinstall)
 binder_call(postinstall, system_server)
 
 # Need to talk to the otadexopt service.
-allow postinstall otadexopt_service:service_manager find;
\ No newline at end of file
+allow postinstall otadexopt_service:service_manager find;
+
+domain_auto_trans(postinstall, otapreopt_chroot_exec, otapreopt_chroot)
+
+# No domain other than update_engine and recovery (via update_engine_sideload)
+# should transition to postinstall, as it is only meant to run during the
+# update.
+neverallow { domain -update_engine -recovery } postinstall:process { transition dyntransition };
diff --git a/postinstall_dexopt.te b/postinstall_dexopt.te
index dbc76df..c5b2533 100644
--- system/sepolicy/postinstall_dexopt.te
+++ system/sepolicy/postinstall_dexopt.te
@@ -8,7 +8,7 @@ type postinstall_dexopt, domain;
 # init_daemon_domain(otapreopt)
 allow postinstall_dexopt self:capability { chown dac_override fowner setgid setuid };
 
-allow postinstall_dexopt postinstall_file:dir getattr;
+allow postinstall_dexopt postinstall_file:dir { getattr search };
 allow postinstall_dexopt proc:file { getattr open read };
 allow postinstall_dexopt tmpfs:file read;
 
@@ -49,9 +49,11 @@ selinux_check_access(postinstall_dexopt)
 # We have to manually transition, as we don't have an entrypoint.
 domain_auto_trans(postinstall_dexopt, postinstall_file, dex2oat)
 
-# installd wants to know about our child.
-allow postinstall_dexopt installd:process sigchld;
+# Postinstall wants to know about our child.
+allow postinstall_dexopt postinstall:process sigchld;
 
 # Allow otapreopt to use file descriptors from otapreopt_chroot.
 # TODO: Probably we can actually close file descriptors...
 allow postinstall_dexopt otapreopt_chroot:fd use;
+
+allow postinstall_dexopt cpuctl_device:dir search;
diff --git a/preopt2cachename.te b/preopt2cachename.te
new file mode 100644
index 0000000..49df647
--- /dev/null
+++ system/sepolicy/preopt2cachename.te
@@ -0,0 +1,13 @@
+# preopt2cachename executable
+#
+# This executable translates names from the preopted versions the build system
+# creates to the names the runtime expects in the data directory.
+type preopt2cachename, domain;
+type preopt2cachename_exec, exec_type, file_type;
+
+# Allow write to stdout.
+allow preopt2cachename cppreopts:fd use;
+allow preopt2cachename cppreopts:fifo_file { getattr read write };
+
+# Allow write to logcat.
+allow preopt2cachename proc_net:file r_file_perms;
diff --git a/priv_app.te b/priv_app.te
index d380a67..85516a6 100644
--- system/sepolicy/priv_app.te
+++ system/sepolicy/priv_app.te
@@ -45,6 +45,10 @@ allow priv_app mnt_media_rw_file:dir search;
 allow priv_app { cache_file cache_recovery_file }:dir create_dir_perms;
 allow priv_app { cache_file cache_recovery_file }:file create_file_perms;
 
+# Write to /data/ota_package for OTA packages.
+allow priv_app ota_package_file:dir rw_dir_perms;
+allow priv_app ota_package_file:file create_file_perms;
+
 # Access to /data/media.
 allow priv_app media_rw_data_file:dir create_dir_perms;
 allow priv_app media_rw_data_file:file create_file_perms;
diff --git a/property.te b/property.te
index 83208cf..af7013f 100644
--- system/sepolicy/property.te
+++ system/sepolicy/property.te
@@ -25,12 +25,14 @@ type audio_prop, property_type, core_property_type;
 type log_prop, property_type, log_property_type;
 type log_tag_prop, property_type, log_property_type;
 type logd_prop, property_type, core_property_type;
+type logpersistd_logging_prop, property_type;
 type mmc_prop, property_type;
 type restorecon_prop, property_type, core_property_type;
 type security_prop, property_type, core_property_type;
 type bluetooth_prop, property_type, core_property_type;
 type pan_result_prop, property_type, core_property_type;
 type powerctl_prop, property_type, core_property_type;
+type cppreopt_prop, property_type, core_property_type;
 type nfc_prop, property_type, core_property_type;
 type dalvik_prop, property_type, core_property_type;
 type config_prop, property_type, core_property_type;
diff --git a/property_contexts b/property_contexts
index dacf2cb..4368a98 100644
--- system/sepolicy/property_contexts
+++ system/sepolicy/property_contexts
@@ -21,6 +21,7 @@ ro.runtime.             u:object_r:system_prop:s0
 hw.                     u:object_r:system_prop:s0
 ro.hw.                  u:object_r:system_prop:s0
 sys.                    u:object_r:system_prop:s0
+sys.cppreopt            u:object_r:cppreopt_prop:s0
 sys.powerctl            u:object_r:powerctl_prop:s0
 sys.usb.ffs.            u:object_r:ffs_prop:s0
 service.                u:object_r:system_prop:s0
@@ -43,6 +44,8 @@ persist.audio.          u:object_r:audio_prop:s0
 persist.debug.          u:object_r:persist_debug_prop:s0
 persist.logd.           u:object_r:logd_prop:s0
 persist.logd.security   u:object_r:device_logging_prop:s0
+persist.logd.logpersistd        u:object_r:logpersistd_logging_prop:s0
+logd.logpersistd        u:object_r:logpersistd_logging_prop:s0
 persist.log.tag         u:object_r:log_tag_prop:s0
 persist.mmc.            u:object_r:mmc_prop:s0
 persist.sys.            u:object_r:system_prop:s0
diff --git a/recovery.te b/recovery.te
index d5767ed..209a276 100644
--- system/sepolicy/recovery.te
+++ system/sepolicy/recovery.te
@@ -7,6 +7,9 @@ type recovery, domain, domain_deprecated;
 # But the allow rules are only included in the recovery policy.
 # Otherwise recovery is only allowed the domain rules.
 recovery_only(`
+  # Allow recovery to perform an update as update_engine would do.
+  typeattribute recovery update_engine_common, boot_control_hal;
+
   allow recovery self:capability { chown dac_override fowner fsetid setfcap setuid setgid sys_admin sys_tty_config };
 
   # Set security contexts on files that are not known to the loaded policy.
diff --git a/service.te b/service.te
index 6b5838c..6f9ab3f 100644
--- system/sepolicy/service.te
+++ system/sepolicy/service.te
@@ -39,6 +39,7 @@ type contexthub_service, app_api_service, system_server_service, service_manager
 type IProxyService_service, app_api_service, system_server_service, service_manager_type;
 type commontime_management_service, system_server_service, service_manager_type;
 type connectivity_service, app_api_service, system_server_service, service_manager_type;
+type connmetrics_service, app_api_service, system_server_service, service_manager_type;
 type consumer_ir_service, app_api_service, system_server_service, service_manager_type;
 type content_service, app_api_service, system_server_service, service_manager_type;
 type country_detector_service, app_api_service, system_server_service, service_manager_type;
@@ -80,6 +81,7 @@ type network_score_service, system_api_service, system_server_service, service_m
 type network_time_update_service, system_server_service, service_manager_type;
 type notification_service, app_api_service, system_server_service, service_manager_type;
 type otadexopt_service, system_server_service, service_manager_type;
+type overlay_service, app_api_service, system_server_service, service_manager_type;
 type package_service, app_api_service, system_server_service, service_manager_type;
 type permission_service, app_api_service, system_server_service, service_manager_type;
 type persistent_data_block_service, system_api_service, system_server_service, service_manager_type;
diff --git a/service_contexts b/service_contexts
index 0ddbdc1..b7d7473 100644
--- system/sepolicy/service_contexts
+++ system/sepolicy/service_contexts
@@ -19,9 +19,10 @@ carrier_config                            u:object_r:radio_service:s0
 clipboard                                 u:object_r:clipboard_service:s0
 com.android.net.IProxyService             u:object_r:IProxyService_service:s0
 commontime_management                     u:object_r:commontime_management_service:s0
-common_time.clock                        u:object_r:mediaserver_service:s0
-common_time.config                       u:object_r:mediaserver_service:s0
+common_time.clock                         u:object_r:mediaserver_service:s0
+common_time.config                        u:object_r:mediaserver_service:s0
 connectivity                              u:object_r:connectivity_service:s0
+connmetrics                               u:object_r:connmetrics_service:s0
 consumer_ir                               u:object_r:consumer_ir_service:s0
 content                                   u:object_r:content_service:s0
 contexthub_service                        u:object_r:contexthub_service:s0
@@ -92,6 +93,7 @@ network_time_update_service               u:object_r:network_time_update_service
 nfc                                       u:object_r:nfc_service:s0
 notification                              u:object_r:notification_service:s0
 otadexopt                                 u:object_r:otadexopt_service:s0
+overlay                                   u:object_r:overlay_service:s0
 package                                   u:object_r:package_service:s0
 permission                                u:object_r:permission_service:s0
 persistent_data_block                     u:object_r:persistent_data_block_service:s0
diff --git a/shell.te b/shell.te
index 056eeb6..3e95b46 100644
--- system/sepolicy/shell.te
+++ system/sepolicy/shell.te
@@ -70,6 +70,7 @@ set_prop(shell, powerctl_prop)
 set_prop(shell, log_tag_prop)
 set_prop(shell, wifi_log_prop)
 userdebug_or_eng(`set_prop(shell, log_prop)')
+userdebug_or_eng(`set_prop(shell, logpersistd_logging_prop)')
 
 # systrace support - allow atrace to run
 allow shell debugfs_tracing:dir r_dir_perms;
diff --git a/surfaceflinger.te b/surfaceflinger.te
index 38f1dad..7364e5f 100644
--- system/sepolicy/surfaceflinger.te
+++ system/sepolicy/surfaceflinger.te
@@ -60,6 +60,9 @@ allow surfaceflinger gpu_service:service_manager { add find };
 allow surfaceflinger surfaceflinger_service:service_manager { add find };
 allow surfaceflinger window_service:service_manager find;
 
+# allow self to set SCHED_FIFO
+allow surfaceflinger self:capability sys_nice;
+
 ###
 ### Neverallow rules
 ###
diff --git a/system_app.te b/system_app.te
index 2d51c5a..50320c5 100644
--- system/sepolicy/system_app.te
+++ system/sepolicy/system_app.te
@@ -32,6 +32,7 @@ set_prop(system_app, logd_prop)
 set_prop(system_app, net_radio_prop)
 set_prop(system_app, system_radio_prop)
 set_prop(system_app, log_tag_prop)
+userdebug_or_eng(`set_prop(system_app, logpersistd_logging_prop)')
 auditallow system_app net_radio_prop:property_service set;
 auditallow system_app system_radio_prop:property_service set;
 
diff --git a/system_server.te b/system_server.te
index e74f58c..3ca8182 100644
--- system/sepolicy/system_server.te
+++ system/sepolicy/system_server.te
@@ -89,6 +89,7 @@ allow system_server { appdomain autoplay_app }:process { getsched setsched };
 allow system_server audioserver:process { getsched setsched };
 allow system_server cameraserver:process { getsched setsched };
 allow system_server mediaserver:process { getsched setsched };
+allow system_server bootanim:process { getsched setsched };
 
 # Read /proc/pid data for all domains. This is used by ProcessCpuTracker
 # within system_server to keep track of memory and CPU usage for
@@ -347,6 +348,9 @@ userdebug_or_eng(`set_prop(system_server, wifi_log_prop)')
 set_prop(system_server, ctl_default_prop)
 set_prop(system_server, ctl_bugreport_prop)
 
+# cppreopt property
+set_prop(system_server, cppreopt_prop)
+
 # Create a socket for receiving info from wpa.
 type_transition system_server wifi_data_file:sock_file system_wpa_socket;
 type_transition system_server wpa_socket:sock_file system_wpa_socket;
@@ -437,6 +441,7 @@ allow system_server mediacodec_service:service_manager find;
 allow system_server mediadrmserver_service:service_manager find;
 allow system_server netd_service:service_manager find;
 allow system_server nfc_service:service_manager find;
+allow system_server overlay_service:service_manager find;
 allow system_server radio_service:service_manager find;
 allow system_server system_server_service:service_manager { add find };
 allow system_server surfaceflinger_service:service_manager find;
@@ -534,7 +539,7 @@ allow system_server update_engine:fifo_file write;
 
 # Access to /data/preloads
 allow system_server preloads_data_file:file { r_file_perms unlink };
-allow system_server preloads_data_file:dir { r_dir_perms write remove_name };
+allow system_server preloads_data_file:dir { r_dir_perms write remove_name rmdir };
 
 ###
 ### Neverallow rules
diff --git a/uncrypt.te b/uncrypt.te
index c8840dd..2d95b88 100644
--- system/sepolicy/uncrypt.te
+++ system/sepolicy/uncrypt.te
@@ -19,6 +19,10 @@ userdebug_or_eng(`
 allow uncrypt cache_recovery_file:dir rw_dir_perms;
 allow uncrypt cache_recovery_file:file create_file_perms;
 
+# Read OTA zip file at /data/ota_package/.
+allow uncrypt ota_package_file:dir r_dir_perms;
+allow uncrypt ota_package_file:file r_file_perms;
+
 # Write to /dev/socket/uncrypt
 unix_socket_connect(uncrypt, uncrypt, uncrypt)
 
diff --git a/untrusted_app.te b/untrusted_app.te
index 6bc6843..35c811c 100644
--- system/sepolicy/untrusted_app.te
+++ system/sepolicy/untrusted_app.te
@@ -31,6 +31,7 @@ allow untrusted_app app_data_file:file { rx_file_perms execmod };
 
 # ASEC
 allow untrusted_app asec_apk_file:file r_file_perms;
+allow untrusted_app asec_apk_file:dir r_dir_perms;
 # Execute libs in asec containers.
 allow untrusted_app asec_public_file:file { execute execmod };
 
@@ -105,6 +106,10 @@ r_dir_file(untrusted_app, proc_net)
 allow untrusted_app sysfs_hwrandom:dir search;
 allow untrusted_app sysfs_hwrandom:file r_file_perms;
 
+# Allow apps to view preloaded content
+allow untrusted_app preloads_data_file:dir r_dir_perms;
+allow untrusted_app preloads_data_file:file r_file_perms;
+
 ###
 ### neverallow rules
 ###
@@ -182,6 +187,7 @@ neverallow untrusted_app {
   fs_type
   -fuse                     # sdcard
   -sdcardfs                 # sdcard
+  -vfat
   file_type
   -app_data_file            # The apps sandbox itself
   -media_rw_data_file       # Internal storage. Known that apps can
diff --git a/update_engine.te b/update_engine.te
index 33e8134..fa3f05c 100644
--- system/sepolicy/update_engine.te
+++ system/sepolicy/update_engine.te
@@ -1,6 +1,6 @@
 # Domain for update_engine daemon.
 # update_engine uses the boot_control_hal.
-type update_engine, domain, domain_deprecated, boot_control_hal;
+type update_engine, domain, domain_deprecated, update_engine_common, boot_control_hal;
 type update_engine_exec, exec_type, file_type;
 type update_engine_data_file, file_type, data_file_type;
 
@@ -21,38 +21,16 @@ dontaudit update_engine kernel:process setsched;
 allow update_engine update_engine_data_file:dir { create_dir_perms };
 allow update_engine update_engine_data_file:file { create_file_perms };
 
-# Allow update_engine to reach block devices in /dev/block.
-allow update_engine block_device:dir search;
-
-# Allow read/write on system and boot partitions.
-allow update_engine boot_block_device:blk_file rw_file_perms;
-allow update_engine system_block_device:blk_file rw_file_perms;
-
 # Don't allow kernel module loading, just silence the logs.
 dontaudit update_engine kernel:system module_request;
 
-# Allow update_engine to mount on the /postinstall directory and reset the
-# labels on the mounted filesystem to postinstall_file.
-allow update_engine postinstall_mnt_dir:dir mounton;
-allow update_engine postinstall_file:filesystem { mount unmount relabelfrom relabelto };
-allow update_engine labeledfs:filesystem relabelfrom;
-
-# Allow update_engine to read and execute postinstall_file.
-allow update_engine postinstall_file:file rx_file_perms;
-allow update_engine postinstall_file:lnk_file r_file_perms;
-allow update_engine postinstall_file:dir r_dir_perms;
-
-# The postinstall program is run by update_engine and will always be tagged as a
-# postinstall_file regardless of its attributes in the new system.
-domain_auto_trans(update_engine, postinstall_file, postinstall)
-
-# A postinstall program is typically a shell script (with a #!), so we allow
-# to execute those.
-allow update_engine shell_exec:file rx_file_perms;
-
 # Register the service to perform Binder IPC.
 binder_use(update_engine)
 allow update_engine update_engine_service:service_manager { add };
 
 # Allow update_engine to call the callback function provided by priv_app.
 binder_call(update_engine, priv_app)
+
+# Read OTA zip file at /data/ota_package/.
+allow update_engine ota_package_file:file r_file_perms;
+allow update_engine ota_package_file:dir r_dir_perms;
diff --git a/update_engine_common.te b/update_engine_common.te
new file mode 100644
index 0000000..e70e44d
--- /dev/null
+++ system/sepolicy/update_engine_common.te
@@ -0,0 +1,37 @@
+# update_engine payload application permissions. These are shared between the
+# background daemon and the recovery tool to sideload an update.
+
+# Allow update_engine to reach block devices in /dev/block.
+allow update_engine_common block_device:dir search;
+
+# Allow read/write on system and boot partitions.
+allow update_engine_common boot_block_device:blk_file rw_file_perms;
+allow update_engine_common system_block_device:blk_file rw_file_perms;
+
+# Allow to set recovery options in the BCB. Used to trigger factory reset when
+# the update to an older version (channel change) or incompatible version
+# requires it.
+allow update_engine_common misc_block_device:blk_file rw_file_perms;
+
+# Allow update_engine_common to mount on the /postinstall directory and reset the
+# labels on the mounted filesystem to postinstall_file.
+allow update_engine_common postinstall_mnt_dir:dir mounton;
+allow update_engine_common postinstall_file:filesystem { mount unmount relabelfrom relabelto };
+allow update_engine_common labeledfs:filesystem relabelfrom;
+
+# Allow update_engine_common to read and execute postinstall_file.
+allow update_engine_common postinstall_file:file rx_file_perms;
+allow update_engine_common postinstall_file:lnk_file r_file_perms;
+allow update_engine_common postinstall_file:dir r_dir_perms;
+
+# The postinstall program is run by update_engine_common and will always be tagged as a
+# postinstall_file regardless of its attributes in the new system.
+domain_auto_trans(update_engine_common, postinstall_file, postinstall)
+
+# A postinstall program is typically a shell script (with a #!), so we allow
+# to execute those.
+allow update_engine_common shell_exec:file rx_file_perms;
+
+# Allow update_engine_common to suspend, resume and kill the postinstall program.
+allow update_engine_common postinstall:process { signal sigstop };
+
diff --git a/update_verifier.te b/update_verifier.te
index 65438d3..09d5fc4 100644
--- system/sepolicy/update_verifier.te
+++ system/sepolicy/update_verifier.te
@@ -5,4 +5,13 @@ type update_verifier_exec, exec_type, file_type;
 
 init_daemon_domain(update_verifier)
 
-# TODO: Add rules to allow update_verifier to read system_block_device.
+# Allow update_verifier to reach block devices in /dev/block.
+allow update_verifier block_device:dir search;
+
+# Read care map in /data/ota_package/.
+allow update_verifier ota_package_file:dir r_dir_perms;
+allow update_verifier ota_package_file:file r_file_perms;
+
+# Read all blocks in system partition.
+allow update_verifier system_block_device:blk_file r_file_perms;
+
diff --git a/zygote.te b/zygote.te
index 89dccfc..c6b343c 100644
--- system/sepolicy/zygote.te
+++ system/sepolicy/zygote.te
@@ -39,6 +39,12 @@ allow zygote dex2oat_exec:file rx_file_perms;
 # Control cgroups.
 allow zygote cgroup:dir create_dir_perms;
 allow zygote self:capability sys_admin;
+# Allow zygote to stat the files that it opens. The zygote must
+# be able to inspect them so that it can reopen them on fork
+# if necessary: b/30963384
+allow zygote pmsg_device:chr_file { getattr };
+allow zygote debugfs_trace_marker:file { getattr };
+
 # Check validity of SELinux context before use.
 selinux_check_context(zygote)
 # Check SELinux permissions.
@@ -82,31 +88,6 @@ userdebug_or_eng(`
 ')
 
 ###
-### A/B OTA
-###
-
-# The zygote is responsible for detecting A/B OTA artifacts and moving them into
-# the actual dalvik-cache.
-
-# Allow zygote access to files in /data/ota.
-# This includes reading symlinks in /data/ota/dalvik-cache. This is required for PIC mode boot
-# images, where the oat file is symlinked to the original file in /system.
-r_dir_file(zygote, ota_data_file)
-
-# The zygote renames the OTA dalvik-cache to the regular dalvik-cache.
-allow zygote ota_data_file:dir { rw_dir_perms rename reparent };
-
-# And needs to relabel the entries, so as to have the dalvikcache_data_file label.
-allow zygote ota_data_file:{ dir file lnk_file } relabelfrom;
-allow zygote dalvikcache_data_file:{ dir file lnk_file } relabelto;
-
-# The zygote also cleans up the now-empty dalvik-cache directory after an OTA.
-# In case something goes wrong in relabelling, we also need to be able to delete the files that
-# have already been moved.
-allow zygote ota_data_file:dir rmdir;
-allow zygote ota_data_file:{ file lnk_file } unlink;
-
-###
 ### neverallow rules
 ###
 
