--- packages/apps/unifiedNlpBackend/Android.mk	2016-11-19 11:01:18.520245433 +1100
+++ packages/apps/unifiedNlpBackend/Android.mk	2016-11-19 11:07:11.764260367 +1100
@@ -11,11 +11,11 @@
 unifiednlp_build := $(unifiednlp_root)/build
 unifiednlp_apk   := build/outputs/apk/unifiedNlpBackend-release-unsigned.apk
 $(unifiednlp_root)/$(unifiednlp_apk):
-        rm -Rf $(unifiednlp_build)
-        mkdir -p $(unifiednlp_out)
-        mkdir -p $(unifiednlp_build)
-        ln -sf $(unifiednlp_out) $(unifiednlp_build)
-        cd $(unifiednlp_root) && JAVA_TOOL_OPTIONS="$(JAVA_TOOL_OPTIONS) -Dfile.encoding=UTF8" gradle assembleRelease
+	rm -Rf $(unifiednlp_build)
+	mkdir -p $(unifiednlp_out)
+	mkdir -p $(unifiednlp_build)
+	ln -sf $(unifiednlp_out) $(unifiednlp_build)
+	cd $(unifiednlp_root) && JAVA_TOOL_OPTIONS="$(JAVA_TOOL_OPTIONS) -Dfile.encoding=UTF8" gradle assembleRelease
 
 LOCAL_CERTIFICATE := platform
 LOCAL_SRC_FILES := $(unifiednlp_apk)
